README.md
# Site Categories

**INACTIVE NOTICE: This plugin is unsupported by WPMUDEV, we've published it here for those technical types who might want to fork and maintain it for their needs.**

## Translations

Translation files can be found at https://github.com/wpmudev/translations

## Organize sites into categories and make it easier for users to search and discover new content on your network with Site Categories.

![site-735x470](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-735x470-583x372.jpg)

 Quickly add site categories to your network.

### Super-Charge Core Structure

Site Categories adds a new level of flexibility and control to Multisite creating a more connected network experience for your users. Perfect for an e-commerce mall, education network or blog community.

### Make Your Network More Networky

Give site administrators the ability to select related categories at signup. Create as many categories as you need to best organize your network. Create a simpler more beautiful navigation experience with site descriptions and avatars. 

### Display Options

Built-in List, Grid, Dropdown and Accordion style settings let you fit any style. Both unordered and order options, provide control and versatility for designing the best layout for your network.

![settings-735x470](https://premium.wpmudev.org/wp-content/uploads/2012/07/settings-735x470-583x372.jpg)

 Simple configuration for any style layout.

 Set category page display options with custom category icons, category descriptions, sites per page and how page titles display with a custom prefix.

### 3 Included Widgets

This plugin comes packaged with three widgets for natural integration.

*   Site categories list
*   Categories cloud widget
*   List sites from a single category

### A Better Multisite

Create a more searchable network and give your users more control. Categories expand the power of Multisite by making sites easier to discover.

## Usage

### To Get Started:

Start by reading the [Installing Plugins](https://premium.wpmudev.org/manuals/wpmu-manual-2/installing-regular-plugins-on-wpmu/) section in our comprehensive [WordPress and WordPress Multisite Manual](https://premium.wpmudev.org/wpmu-manual/) if you are new to WordPress.

### To Install:

1.  Download the plugin file.
2.  Unzip the file into a folder on your hard drive.
3.  Upload the **_/site-categories/_** folder and all its contents to the **/wp-content/plugins/** folder on your site.
4.  Login to your network admin panel and network-activate the plugin:

_Please note:_ Once the plugin is installed and network-activated, you'll see a new menu item in the dashboard of your main site called Site Categories. This is where you will set the options for the plugin. There are NO network options for this plugin. 

![site-categories-main-settings-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-main-settings-1078.png)

 Let's begin by checking out the new **Site Categories** on the main site. This will help you preview your layout options in the following steps.

#### Creating Your Site Categories

Click the "Site Categories" sub-menu item in the admin area of your main site. You may recognize the layout here. It's your run-of-the-mill add categories menu, only this time it's for your site categories. 

![site-categories-add-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-add-1078.png)

 In our examples on this page, we've created categories for a network of specialized stores. Go ahead and add a few categories that fit the type of sites in your network now (you can easily edit these later if you wish). You may also have noticed the Icon column here. That enables you to optionally upload an icon for each site category after you create it. Simply hover your mouse pointer over any category and click the Edit link to upload your icon. 

![site-categories-icon-add-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-icon-add-1078.png)

 If you don't want a specific icon for a site category, simply leave it blank to use your default image. We'll get to that in a minute. And that's all for that part! Now lets go to "Site Categories" > "Settings" and look around in there.

#### Landing Page Selection

This first section lets you set the Landing Page. Simply select an existing page on your main site where you want the Site Categories listing to be displayed. If you want to create a new page for this, go do that now. Then come back and select it from the dropdown here. 

![site-categories-1080-landing-page](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-1080-landing-page.png)

 You can also choose to use Pretty URLs so that they look like this: `http://yoursite.com/sites/magazines-newspapers/` ...instead of this: `http://yoursite.com/sites/?category=magazines-newspapers`

#### Site Categories Selection

Next up: the Site Categories Selection Options. This section enables you to set exactly what your site admins are able to select when they are configuring the categories for their site. 

![1\. Enter how many categories site admins can select. 2\. Adjust the numbers for Pro Sites sites. 3\. Exclude some categories from selection if you wish. 4\. Enable or disable parent category selection.](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-selection-options-1078.png)

 1\. Enter how many categories site admins can select.  
2\. Adjust the numbers for Pro Sites sites.  
3\. Exclude some categories from selection if you wish.  
4\. Enable or disable parent category selection.

 The _Number of Categories per Site_ lets your site admins pick more than one category, but you can control how many total with this setting. The _Pro Site_ option that you see here will only show if you have Pro Sites installed. It lets you restrict the number of categories a Pro Sites site admin can choose. Note: There are no additional options in the Pro Sites menu, this is the only setting for Pro Site action. The _Excluded Categories_ option is useful if you want to reserve some site categories that only you, as network admin, have access to. _Category Parents Selectable_ simply allows or disallows the selection of parent categories by site admins.

#### Landing Page Categories Display

This is where you begin to actually set things up the way you want them to display on the front-end of your main site. The first thing you'll notice is that the available options will change depending on what you select in the _Display Style_ dropdown. 

![site-categories-landing-page-1-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-landing-page-1-1078.png)

 The Unordered List and Ordered List options output simple vertical lists of your store categories. The Accordion and Grid options output, well, accordion or grid layouts. Here's what the grid and accordion layouts might look like on our example site. 

![site-categories-1-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-1-1078.png)

 

![site-categories-2-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-2-1078.png)

 We'll only go over the options for a grid layout here. Then you can have fun exploring how things work with the other layout styles.

![1\. Enter the number of rows and columns for your grid. 2\. Select how they should be ordered on the page. 3\. Select whether you want empty categories to display or not. 4\. Select to show the counts for parent and child categories. 5\. Enable/disable the display of category descriptions. 6\. Enable/disable the display of category icons. 7\. Enter your preferred icon size. 8\. Upload a default category icon if you wish.](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-landing-page-2-10781.png)

 1\. Enter the number of rows and columns for your grid.  
2\. Select how they should be ordered on the page.  
3\. Select whether you want empty categories to display or not.  
4\. Select to show the counts for parent and child categories.  
5\. Enable/disable the display of category descriptions.  
6\. Enable/disable the display of category icons.  
7\. Enter your preferred icon size.  
8\. Upload a default category icon if you wish.

 The _Categories per Page_ option allows you to set up your grid display exactly the way you want it. Note that the grid display is _not_ responsive. If you are using a responsive theme, you may wish to consider using a list or accordion layout. The _Order By_ option enables you to set the order by category Name or ID, or none at all. You can also select to have them display in ascending or descending order (alphabetically by Name, or numerically by ID). _Hide Empty Categories_ will, if enabled, simply prevent the display of site categories that don't have any sites in them yet. If you set the _Show Counts_ option to Yes, the number of sites in each category will display beside the category name. Enable _Show Category Description_ to display the descriptions you entered when creating or editing your site categories (note that for clarity in our screenshots, we have elected to not display the category descriptions). The _Show Icons_ and _Icon Size_ settings simply control the optional display of the category icons you may have added to each category. The _Default Category Image_ setting is where you can upload an image that will be used as the icon for any category that doesn't have its own. Remember we left some blank earlier when creating our site categories? This is where we upload the icon to use for those ones.

#### Landing Page Sites Display

This section of the settings is where you set up how you want individual category pages to display. These settings should be familiar by now, so have a go and play around with them. 

![site-categories-landing-page-sites-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-landing-page-sites-1078.png)

 Below is an example of how a site page could look. We've chosen to display site descriptions here so you can see how they display. 

![site-categories-landing-page-sites-2-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-landing-page-sites-2-1078.png)

 You may have also noticed that, in our screenshot, individual sites have an avatar! This requires our [Avatars](https://premium.wpmudev.org/project/avatars/ "WordPress Avatars Plugin - WPMU DEV") plugin. If you do not have that plugin installed, you'll see a friendly reminder instead of the settings.

#### New Site Signup Form

Finally, we have the settings for the form that you can choose to have displayed when users register on your network. 

![site-categories-signup-form-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-signup-form-1078.png)

 _Show Site Categories section_ turns the form on or off on your site's signup or register page. _Site Categories Selection Required_ should be set to "Yes" if you want category selection to be mandatory when users create new sites in your network. _Label for Site Categories Dropdowns_ is where you can enter the text you want to display above the category selection dropdowns. Set _Description is Required_ to "Yes" is you want site descriptions to be mandatory when users create new sites. _Label for Site Categories Description_ is where you can enter the text you want to display above the site description text area. Here's what the signup form would look like using our example site categories: 

![site-categories-signup-form-2-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-signup-form-2-1078.png)

 Note that the number of category dropdowns that appear here depends on the number you had set in the _Site Categories Selection_ settings. Also, even if Pro Sites is active on your network, the number of selectable categories you had set for Pro Sites levels will **not** display here. That's because the site hasn't been created yet, and the plugin cannot yet know what Pro Site level the site is at. And that's it for the main site settings. Now let's take a look at what your site admins see in their dashboards.

#### Sub-site Admin Settings

The sub-sites will all have a new menu option under the **Settings** tab called _Site Categories_. 

![site-categories-subsite-settings-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-subsite-settings-1078.png)

 In that admin area, your sub-site admins can select or modify the categories they want their sites to appear under. 

![site-categories-subsite-1078](https://premium.wpmudev.org/wp-content/uploads/2012/07/site-categories-subsite-1078.png)


changelog.txt
----------------------------------------------------------------------
Site Categories
Author: Paul Menard (Incsub), Ashok, Daniel
----------------------------------------------------------------------

1.0.9.4 - 2017-08-31
----------------------------------------------------------------------
- Fixed: Wrong link being displayed on view link in admin.
- Fixed: Prevent various PHP errors.
- Fixed: Typos.
- Fixed: Only show category specific sites on single category widget.

1.0.9.3 - 2017-05-31
----------------------------------------------------------------------
- Fixed: The sorting of site categories on signup page
- Fixed: Site description not being added for new blogs
- Fixed: Default category not added when none selected from signup page

1.0.9.2 - 2017-05-30
----------------------------------------------------------------------
- Fixed: Image Upload Button Does Not Work
- Fixed: "Category parents selectable" is set to "no" hides child categories
- Fixed: PHP Notice for Site Categories outdated method
- Fixed: Clicking on number of sites was generating wrong URL

1.0.9.1 - 2014-03-02
----------------------------------------------------------------------
- Corrected reported issue where empty category items on the landing page were improperly linked. 

1.0.9 - 2014-02-12
----------------------------------------------------------------------
- Added option on accordion output to link to parent category.
- Added option to open site links in new window for landing page as well as all three widgets. 
- Added option to show return link on sites listing to return to main landing page listing.
- Added landing page output option for nested unordered list, nested ordered list and nested select.
- Added accordion display option for child categories to show in flat unordered list, nested unordered list, flat ordered list, nested ordered list OL flat select and nested select.
- Added grid display option for child categories to show in flat unordered list, nested unordered list, flat ordered list, nested ordered list OL flat select and nested select.


1.0.8.8 - 2014-01-09
----------------------------------------------------------------------
- Corrections to WP Media popup loading. Corrections to loading the new media libraries used for WP 3.5.
- Corrections to Accordion landing page display where zero count categories were improperly being linked. 
- Corrections to content formatting for sub-category desctiptions. Previously calling content filters which causes recursive looping. 
- Corrections to inherited Walker class showing PHP5 strict errors.

1.0.8.7 - 2013-12-20
----------------------------------------------------------------------
- Updated avatar selection buttons to use new WP Media popup over thickbox.

1.0.8.6 - 2013-12-12
----------------------------------------------------------------------
- Corrected issues enqueueing JS/CSS to be compatable with https in dashboard.
- Ensure all avatars and icons on dashboard and front-end are being set to https:// when SSL used.
- Added logic to prevent loading plugin if installed on non-Multisite WordPress site.

1.0.8.5 - 2013-11-25
----------------------------------------------------------------------
- Added Include 'AND' option to Site from single site categoy widget.
- Corrected logic for the include OR/AND category IDs 
- Check UI works with new WP 38 and TwentyFourteen theme
- Updates to new WPMU DEV Dashboard notification library

1.0.8.4 - 2013-10-31
----------------------------------------------------------------------
- Added Include/Exclude/Exclude Tree option for Site Category IDs to all three widgets
- Added Include/Exclude option for Blog IDs to Category Sites widget
- Added Output Option Select Nested/Flat option categories widget
- Corrected issue on widget admin to show dropdown categories in alph order

1.0.8.3 - 2013-10-16
----------------------------------------------------------------------
- Corrected incorrect version number '1.0.8.2-Beta1' in previous update. should have been '1.0.8.2'. Causing update issues for members.

1.0.8.2 - 2013-10-16
----------------------------------------------------------------------
- Corrected issue on new filter added in previous version. Now returning correct value.
- Corrected issue on new dropdown display option

1.0.8.1 - 2013-10-08
----------------------------------------------------------------------
- Added options for Widget Site Categories to display output as select.
- Added options for Widget Site Categories Sites to display output as select.
- Added support for BuddyPress register form.
- Added filter 'site_categories_user_can_view' to allow hiding the landing page or widgets content from being displayed on certain pages or to certain users.
- Fixed reported issue with Cloud widget not showing when showing all categories. 

1.0.8 - 2013-09-02
----------------------------------------------------------------------
- Added option to disable rewrite rules (Pretty URLs) for Category display. Needed if landing page set to home.
- Corrected/Added options to order category site listings.
- Added support for a default virtual category to display all non-assigned sites to help with initial install of plugin. 

1.0.7.9 - 2013-08-20
----------------------------------------------------------------------
- Added filter to remove inner html from images used. 
- Reverted back to using thickbox for category icon selection. Until issues can be resolved with WP new media selector.

1.0.7.8 - 2013-06-19
----------------------------------------------------------------------
- Update media selection from using Thickbox to newer Media popup from WP 3.5

1.0.7.7 - 2013-04-09
----------------------------------------------------------------------
- Updated plugin JavaScript for upcoming WordPress 3.6/jQuery 1.9 changes. 

1.0.7.6 - 2013-03-19 

lib/widgets/class_site_categories_widget_cloud.php
<?php
/**
 * Shows all Blog categories.
 */
class Bcat_WidgetCloud extends WP_Widget {

	function __construct () {
		$widget_ops = array('classname' => __CLASS__, 'description' => __('Shows a tag cloud of Site Categories.', SITE_CATEGORIES_I18N_DOMAIN));
		parent::__construct(__CLASS__, __('Site Categories Cloud', SITE_CATEGORIES_I18N_DOMAIN), $widget_ops);
	}

	function form($instance) {
		global $current_site;
		
		// Set defaults
		// ...
		$defaults = array( 
			'title' 				=> 	'',
			'number'				=>	'',
			'orderby'				=>	'name',
			'order'					=>	'ASC',
			'smallest'				=>	'8',
			'largest'				=>	'22',
			'unit'					=>	'pt',
			'category'				=>	'',
			'category_filter'		=>	'',
			'category_ids'			=>	'',
			'include_parent'		=>	'on',
			'open_blank'			=>	0,
			'show_more_link'		=>	1,
			'landing_link_label'	=>	__('more categories', SITE_CATEGORIES_I18N_DOMAIN)
		);
		
		$instance = wp_parse_args( (array) $instance, $defaults ); 
		//echo "instance<pre>"; print_r($instance); echo "</pre>";
		
		//if (!empty($instance['per_page'])) {
		//	$instance['per_page'] = intval($instance['per_page']);
		//}
		
		?>
		<p>
			<label for="<?php echo $this->get_field_id('title') ?>"><?php _e('Title:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>
			<input type="text" name="<?php echo $this->get_field_name('title'); ?>" id="<?php echo $this->get_field_id('title'); ?>" 
				class="widefat" value="<?php echo $instance['title'] ?> "/>
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'number' ); ?>"><?php 
				_e('Number of item to show: (blank for all)', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<input type="text" id="<?php echo $this->get_field_id( 'number' ); ?>" value="<?php echo $instance['number']; ?>"
				name="<?php echo $this->get_field_name( 'number'); ?>" class="widefat" style="width:100%;" />
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'category' ); ?>"><?php 
				_e('Show Site Categories:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<?php
				switch_to_blog( $current_site->blog_id );
		
				$bcat_args = array(
					'taxonomy'			=> 	SITE_CATEGORIES_TAXONOMY,
					'orderby'			=>	'name',
					'order'				=>	'ASC',
					'hierarchical'		=>	true,
					'hide_empty'		=>	false,
					'show_count'		=>	true,
					'show_option_none'	=>	__('All Categories', SITE_CATEGORIES_I18N_DOMAIN), 
					'name'				=>	$this->get_field_name('category'),
					'class'				=>	'widefat',
					'selected'			=>	$instance['category']
				);
		
				wp_dropdown_categories( $bcat_args ); 
				restore_current_blog();

			?><input type="checkbox" id="<?php echo $this->get_field_id( 'include_parent' ); ?>" <?php if ($instance['include_parent'] == "on") {
				echo ' checked="checked" '; } ?>
				name="<?php echo $this->get_field_name( 'include_parent'); ?>"  /> <label for="<?php echo $this->get_field_id( 'include_parent' ); ?>"><?php 
				_e('Include Parent:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>
		</p>
		<p>
			<label for="<?php echo $this->get_field_id('category_filter') ?>"><?php _e('Site Categories Include/Exclude:', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			<select id="<?php echo $this->get_field_id( 'category_filter' ); ?>" 
				name="<?php echo $this->get_field_name( 'category_filter'); ?>" class="widefat">
				<option value="" <?php if ($instance['category_filter'] == "") { echo ' selected="selected" '; }?>><?php 
					_e('Show All', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="include" <?php if ($instance['category_filter'] == "include") { echo ' selected="selected" '; }?>><?php 
					_e('Include', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="exclude" <?php if ($instance['category_filter'] == "exclude") { echo ' selected="selected" '; }?>><?php 
					_e('Exclude', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="exclude_tree" <?php if ($instance['category_filter'] == "exclude_tree") { echo ' selected="selected" '; }?>><?php 
					_e('Exclude Tree', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
			</select><br />
		</p>
		<p>
			<label for="<?php echo $this->get_field_id('category_ids') ?>"><?php _e('Site Categories IDs (comma seperated):', SITE_CATEGORIES_I18N_DOMAIN); ?></label>
			<input type="text" name="<?php echo $this->get_field_name('category_ids'); ?>" id="<?php echo $this->get_field_id('category_ids'); ?>" 
				class="widefat" value="<?php echo $instance['category_ids'] ?>"/>
		</p>

		<p>
			<label for="<?php echo $this->get_field_id('orderby') ?>"><?php _e('Ordering:', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			<select id="<?php echo $this->get_field_id( 'orderby' ); ?>" 
				name="<?php echo $this->get_field_name( 'orderby'); ?>" class="widefat" style="width:60%;">
				<option value="name" <?php if ($instance['orderby'] == "name") { echo ' selected="selected" '; }?>><?php 
					_e('Name', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="count" <?php if ($instance['orderby'] == "count") { echo ' selected="selected" '; }?>><?php 
					_e('Count', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
			</select>
			<select id="<?php echo $this->get_field_id( 'order' ); ?>" 
				name="<?php echo $this->get_field_name( 'order'); ?>" class="widefat" style="width:25%;">
				<option value="ASC" <?php if ($instance['order'] == "ASC") { echo ' selected="selected" '; }?>><?php 
					_e('ASC', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="DESC" <?php if ($instance['order'] == "DESC") { echo ' selected="selected" '; }?>><?php 
					_e('DESC', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="RAND" <?php if ($instance['order'] == "RAND") { echo ' selected="selected" '; }?>><?php 
					_e('Random', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
			</select>
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'smallest' ); ?>"><?php 
				_e('Small Font size:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<input type="text" id="<?php echo $this->get_field_id( 'smallest' ); ?>" value="<?php echo $instance['smallest']; ?>"
				name="<?php echo $this->get_field_name( 'smallest'); ?>" class="widefat" style="width:100%;" />
		</p>
		<p>
			<label for="<?php echo $this->get_field_id( 'largest' ); ?>"><?php 
				_e('Large Font size:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<input type="text" id="<?php echo $this->get_field_id( 'largest' ); ?>" value="<?php echo $instance['largest']; ?>"
				name="<?php echo $this->get_field_name( 'largest'); ?>" class="widefat" style="width:100%;" />
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'unit' ); ?>"><?php 
				_e('Font size unit:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<select id="<?php echo $this->get_field_id( 'unit' ); ?>" 
				name="<?php echo $this->get_field_name( 'unit'); ?>" class="widefat" style="width:100%;">
				<option value="%" <?php if ($instance['unit'] == "%") { echo ' selected="selected" '; }?>><?php 
					_e('% Percentage', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="in" <?php if ($instance['unit'] == "in") { echo ' selected="selected" '; }?>><?php 
					_e('in Inch', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="cm" <?php if ($instance['unit'] == "cm") { echo ' selected="selected" '; }?>><?php 
					_e('cm Centimeter', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="mm" <?php if ($instance['unit'] == "mm") { echo ' selected="selected" '; }?>><?php 
					_e('mm Millimeter', SITE_CATEGORIES_I18N_DOMAIN); ?></option>					
				<option value="em" <?php if ($instance['unit'] == "em") { echo ' selected="selected" '; }?>><?php 
					_e('em', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="ex" <?php if ($instance['unit'] == "ex") { echo ' selected="selected" '; }?>><?php 
					_e('ex', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="pt" <?php if ($instance['unit'] == "pt") { echo ' selected="selected" '; }?>><?php 
					_e('pt Point', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="pc" <?php if ($instance['unit'] == "pc") { echo ' selected="selected" '; }?>><?php 
					_e('pc Pica', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
				<option value="px" <?php if ($instance['unit'] == "px") { echo ' selected="selected" '; }?>><?php 
					_e('px Pixels', SITE_CATEGORIES_I18N_DOMAIN); ?></option>
			</select>
		</p>
		
		<p>
			<label for="<?php echo $this->get_field_id('open_blank') ?>"><?php _e('Open links in new window:', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			<input type="radio" name="<?php echo $this->get_field_name( 'open_blank'); ?>" id="<?php echo $this->get_field_id('open_blank') ?>_yes" 
				value="1" <?php if ($instance['open_blank'] == "1") { echo ' checked="checked" '; } ?> /> <label for="<?php echo $this->get_field_id('open_blank') ?>_yes"><?php _e('Yes', SITE_CATEGORIES_I18N_DOMAIN); ?></label>
			
			<input type="radio" name="<?php echo $this->get_field_name( 'open_blank'); ?>" id="<?php echo $this->get_field_id('open_blank') ?>_no" 
				value="0" <?php if ($instance['open_blank'] == "0") { echo ' checked="checked" '; } ?> /> <label for="<?php echo $this->get_field_id('open_blank') ?>_no"><?php _e('No', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			
		</p>
		<p>
			<label for="<?php echo $this->get_field_id('show_more_link') ?>"><?php _e('Landing Page link below cloud list:', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			<input type="radio" name="<?php echo $this->get_field_name( 'show_more_link'); ?>" id="<?php echo $this->get_field_id('show_more_link') ?>_yes" 
				value="1" <?php if ($instance['show_more_link'] == "1") { echo ' checked="checked" '; } ?> /> <label for="<?php echo $this->get_field_id('show_more_link') ?>_yes"><?php _e('Yes', SITE_CATEGORIES_I18N_DOMAIN); ?></label>
			
			<input type="radio" name="<?php echo $this->get_field_name( 'show_more_link'); ?>" id="<?php echo $this->get_field_id('show_more_link') ?>_no" 
				value="0" <?php if ($instance['show_more_link'] == "0") { echo ' checked="checked" '; } ?> /> <label for="<?php echo $this->get_field_id('show_more_link') ?>_no"><?php _e('No', SITE_CATEGORIES_I18N_DOMAIN); ?></label><br />
			
		</p>
		
		<p>
			<label for="<?php echo $this->get_field_id( 'landing_link_label' ); ?>"><?php 
				_e('Label for link:', SITE_CATEGORIES_I18N_DOMAIN); ?></label>

			<input type="text" id="<?php echo $this->get_field_id( 'landing_link_label' ); ?>" value="<?php echo $instance['landing_link_label']; ?>"
				name="<?php echo $this->get_field_name( 'landing_link_label'); ?>" class="widefat" style="width:100%;" />
		</p>
		

		<?php
	}

	function update($new_instance, $old_instance) {
		$instance = $old_instance;

		$instance['title'] 					= strip_tags($new_instance['title']);

		if (!empty($new_instance['category']))
			$instance['category'] 			= intval($new_instance['category']);
		else
			$instance['category']			= '';

		if (!empty($new_instance['category_filter']))
			$instance['category_filter'] 	= strip_tags($new_instance['category_filter']);
		else
			$instance['category_filter'] 	= '';
		
		if (!empty($new_instance['category_ids']))
			$instance['category_ids'] 		= strip_tags($new_instance['category_ids']);
		else
			$instance['category_ids'] 		= '';

		if (isset($new_instance['include_parent']))
			$instance['include_parent'] 	= esc_attr($new_instance['include_parent']);
		else
			$instance['include_parent']		= '';


		if (!empty($new_instance['number']))
			$instance['number'] 			= intval($new_instance['number']);
		else
			$instance['number']				= '';
			

		if (isset($new_instance['orderby']))			
			$instance['orderby'] 			= esc_attr($new_instance['orderby']);
		else
			$instance['orderby']			= 'name';

			
		if (isset($new_instance['order']))
			$instance['order'] 			= esc_attr($new_instance['order']);
		else
			$instance['order']			= 'ASC';
			
			
		if (!empty($new_instance['smallest']))
			$instance['smallest'] 		= intval($new_instance['smallest']);
		else
			$instance['smallest']		= 8;


		if (!empty($new_instance['largest']))
			$instance['largest'] 		= intval($new_instance['largest']);
		else
			$instance['largest']		= 22;


		if (!empty($new_instance['unit']))
			$instance['unit'] 			= esc_attr($new_instance['unit']);
		else
			$instance['unit']		= "pt";
		
		$instance['show_more_link'] 	= intval($new_instance['show_more_link']);
		$instance['open_blank'] 		= intval($new_instance['open_blank']);
		$instance['landing_link_label'] = strip_tags($new_instance['landing_link_label']);
		
		delete_site_transient( 'site-categories-cloud-data-'. $this->number );
		return $instance;
	}

	function widget($args, $instance) {

		global $site_categories, $current_site;		
		
		$site_categories->load_config();
		extract($args);

		$data = get_site_transient( 'site-categories-cloud-data-'. $this->number );
		//echo "data<pre>"; print_r($data); echo "</pre>";
		if (!$data) {
			
			switch_to_blog( $current_site->blog_id );

			$defaults = array(
				'smallest' 		=> 	8, 
				'largest' 		=> 	22, 
				'unit' 			=> 	'pt', 
				'number' 		=> 	45,
				'format' 		=> 	'flat', 
				'separator' 	=> 	"\n", 
				'orderby' 		=> 	'name', 
				'order' 		=> 	'ASC',
				'exclude' 		=> 	'', 
				'include' 		=> 	'', 
				'link' 			=> 	'view', 
				'taxonomy' 		=> 	SITE_CATEGORIES_TAXONOMY, 
				'echo' 			=> 	false
			);
			$instance = wp_parse_args( $instance, $defaults );
			if ((isset($instance['category'])) && (intval($instance['category']) > 0)) {
				$instance['child_of'] 	= intval($instance['category']);
			} else {
				$instance['child_of'] 	= 0;
			}
			
			
			if (( isset($instance['category_filter'])) && (!empty($instance['category_filter']))) {
				$instance['category_ids'] = str_replace(' ', '', $instance['category_ids']);
				if ((isset($instance['category_ids'])) && (!empty($instance['category_ids']))) {
					$category_ids = explode( ',', $instance['category_ids'] );
					if (!empty($category_ids)) {
						if ($instance['category_filter'] == 'include') {
							$instance['include'] = $category_ids;
						} else if ($instance['category_filter'] == "exclude") {
							$instance['exclude'] = $category_ids;
					 	} else if ($instance['category_filter'] == 'exclude_tree') {
							$instance['exclude_tree'] = $category_ids;
					 	}
					}
				}
			}
			
			
			//echo "instance<pre>"; print_r($instance); echo "</pre>";
			
			$tags = get_terms( SITE_CATEGORIES_TAXONOMY, $instance); // Always query top tags
			//echo "tags<pre>"; print_r($tags); echo "</pre>";
			if ((isset($instance['include_parent']))	&& ($instance['include_parent'] == "on")) {
				$parent_tag = get_term_by('id', $instance['category'], SITE_CATEGORIES_TAXONOMY);
				if ( !empty( $parent_tag ) && !is_wp_error( $parent_tag ) ) {
					//echo "parent_tag<pre>"; print_r($parent_tag); echo "</pre>";
					$tags[] = $parent_tag;
				}
			}

			if ( empty( $tags ) || is_wp_error( $tags ) )
				return;

			foreach ( $tags as $key => $tag ) {
				
				$tags[ $key ]->id = $tag->term_id;
				if ((isset($site_categories->opts['landing_page_rewrite'])) && ($site_categories->opts['landing_page_rewrite'] == true) && ($site_categories->opts['landing_page_use_rewrite'] == "yes")) {
					$tags[ $key ]->link = trailingslashit($site_categories->opts['landing_page_slug']) . $tag->slug;
				} else {
					//$tags[ $key ]->link = $site_categories->opts['landing_page_slug'] .'&amp;category_name=' . $tag->slug;
					$tags[ $key ]->link = add_query_arg(array('category' => $tag->slug), $site_categories->opts['landing_page_slug']); 
				}
			}
			$data = wp_generate_tag_cloud( $tags, $instance ); // Here's where those top tags get sorted according to $args

			if ($instance['open_blank'] == '1'){
				$data = str_replace('<a ', '<a target="_blank" ', $data);
			}
			
			if (!empty($data)) {
				if ((isset($instance['show_more_link'])) && ($instance['show_more_link'])) { 

					if ($instance['open_blank'] == '1') { $link_target = ' target="_blank" '; } 
					else { $link_target = ''; }
					$data .= '<div id="site-categories-navigation">';
					$data .= '<a href="'. $site_categories->opts['landing_page_slug'] .'" '. $link_target .'>'. $instance['landing_link_label'] .'</a>';
					echo '</div>';
				}
			}

			restore_current_blog();
			
			set_site_transient( 'site-categories-cloud-data-'. $this->number, $data, 30);
		}
		
		$user_access_content = apply_filters('site_categories_user_can_view', '', $this->id_base);
		//echo "user_access_content[". $user_access_content ."]<br />";
		
		// If the filters returned simply false we return the default content'
		if ($user_access_content === false)
			return false;
		
		if ($data) {
			echo $before_widget;

			$title = apply_filters('widget_title', $instance['title']);
			if ($title) echo $before_title . $title . $after_title;
		
			// If the filters returned a string/text we want to use that as the user viewed content 	
			if ((is_string($user_access_content)) && (!empty($user_access_content)))
				echo $user_access_content;
			else
				echo $data;
		
			echo $after_widget;
			
		}
	}
}

